### PR TITLE
Add Typescript declarations

### PR DESCRIPTION
**Pull request by timhc22**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -2,3 +2,4 @@ node_modules
 dist
 yarn-error.log
 *.DS_Store
+.idea
```
Comment (makoto) [inline .gitignore]: What is `.idea` ? Can you remove it?

File: src/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,81 @@
+export default class ENS {
+    constructor(options: any);
+    provider: ethers.providers.Provider;
+    signer: any;
+    ens: any;
+    name(name: any): Name;
```
Comment (makoto) [inline src/index.d.ts]: `name: any` should be `name: string`

File: src/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    resolver(address: any): Resolver;
```
Comment (makoto) [inline src/index.d.ts]: `address: any` should be `address:string`

File: src/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    getName(address: any): Promise<{
```
Comment (makoto) [inline src/index.d.ts]: `address: any` should be `address:string`

File: src/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+        name: any;
+    }>;
+    getNameWithResolver(address: any, resolverAddr: any): Promise<{
```
Comment (makoto) [inline src/index.d.ts]: both `address`  and `resolverAddress` should be `string`

File: src/index.d.ts [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+        name: any;
+    }>;
+    setReverseRecord(name: any, overrides: any): Promise<any>;
```
Comment (makoto) [inline src/index.d.ts]: `name` should be `string`

File: src/index.d.ts [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+}
+import { ethers } from "ethers";
+declare class Name {
+    constructor(options: any);
+    namehash: string;
+    ens: any;
+    ensWithSigner: any;
+    name: any;
```
Comment (makoto) [inline src/index.d.ts]: `name` should be `string`

File: src/index.d.ts [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+    provider: any;
+    signer: any;
+    resolver: any;
+    getOwner(): Promise<any>;
+    setOwner(address: any): Promise<any>;
```
Comment (makoto) [inline src/index.d.ts]: `address: any` should be `address:string`

File: src/index.d.ts [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+    getResolver(): Promise<any>;
+    setResolver(address: any): Promise<any>;
```
Comment (makoto) [inline src/index.d.ts]: `address: any` should be `address:string`

File: src/index.d.ts [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+    getTTL(): Promise<any>;
+    getResolverAddr(): Promise<any>;
+    getAddress(coinId: any): Promise<any>;
```
Comment (makoto) [inline src/index.d.ts]: `coinId` is `number`

File: src/index.d.ts [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+    setAddress(key: any, address: any): Promise<any>;
```
Comment (makoto) [inline src/index.d.ts]: `key` and `address` should be `string`

File: src/index.d.ts [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+        value: any;
+        contentType: string;
+    }>;
+    setContenthash(content: any): Promise<any>;
+    getText(key: any): Promise<any>;
+    setText(key: any, recordValue: any): Promise<any>;
+    setSubnodeOwner(label: any, newOwner: any): Promise<any>;
+    setSubnodeRecord(label: any, newOwner: any, resolver: any, ttl?: number): Promise<any>;
+    createSubdomain(label: any): Promise<any>;
+    deleteSubdomain(label: any): Promise<any>;
```
Comment (makoto) [inline src/index.d.ts]: All `label`, `newOwner`, resolver` and `key` above should be `string`

File: src/index.d.ts [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+}
+declare class Resolver {
+    constructor({ address, ens }: {
+        address: any;
```
Comment (makoto) [inline src/index.d.ts:190]: `address: any` should be `address:string`

File: src/index.d.ts [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+        ens: any;
+    });
+    address: any;
```
Comment (makoto) [inline src/index.d.ts:193]: `address: any` should be `address:string`

File: src/index.d.ts [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+    ens: any;
+    name(name: any): Name;
```
Comment (makoto) [inline src/index.d.ts]: `name: any` should be `name:string`

File: src/index.js [hunk 16]
```diff
@@ -19,10 +19,25 @@ function getEnsAddress(networkId) {
...
+ * @returns {Contract}
+ */
 function getResolverContract({ address, provider }) {
   return new ethers.Contract(address, resolverContract, provider)
 }
 
+/**
+ * Get ENS Contract
+ *
+ * // todo is Provider here meant to be provider from web3-core ? (may need to upgrade web3 library)
```
Comment (makoto) [inline src/index.js]: Save as above

File: src/index.js [hunk 17]
```diff
@@ -19,10 +19,25 @@ function getEnsAddress(networkId) {
   }
 }
 
+/**
+ * Get Resolver Contract
+ * // todo is Provider here meant to be provider from web3-core ? (may need to upgrade web3 library)
```
Comment (makoto) [inline src/index.js]: The `provider` is used at `ethers.Contract` so whatever the type definition ethers.io uses. Ethers.js is written in Type script so probably you can import type from there?

File: src/utils/contents.d.ts [hunk 18]
```diff
@@ -0,0 +1,12 @@
+export function decodeContenthash(encoded: any): {
+    protocolType: any;
```
Comment (makoto) [inline src/utils/contents.d.ts]: string

**Comment by makoto** [comment]
Do we need any instruction on README on how to use this type definitions?

**Comment by martriay** [comment]
Why was this closed?

**Comment by timhc22** [comment]
@martriay, I made an error, it is still open, but here:

https://github.com/ensdomains/ensjs/pull/58

**Pull request closed without merging**